File: aoc-2020/day-01/day_1.py
with open("input.txt") as f:
    expenses = [int(line) for line in f]


def part1():
    pairs = set()
    for n in expenses:
        if n in pairs:
            print(f"Part 1: {n * (2020 - n)}")
            return
        pairs.add(2020 - n)


def part2():
    triplets = {}
    for i, n in enumerate(expenses):
        for m in expenses[i + 1 :]:
            if m in triplets:
                a, b = triplets[m]
                print(f"Part 2: {a*b*m}")
                return
            triplets[2020 - n - m] = (n, m)


part1()
part2()

File: aoc-2015/day-01/day_1.hs
-- stack --resolver lts-16.27 script 
main :: IO ()
main = do
  content <- readFile "input.txt"
  putStrLn $ "Part 1: " ++ show (part1 content)
  putStrLn $ "Part 2: " ++ show (part2 content)

translated :: String -> [Int]
translated s = map translate s
 where
  translate x | x == '('  = 1
              | otherwise = -1

part1 :: String -> Int
part1 s = (sum . translated) s

part2 :: String -> Int
part2 s = index
 where
  (index, _) = head $ filter in_basement (zip [1 ..] cumsum)
  cumsum     = scanl (+) 0 $ translated s
  in_basement (_, level) = level < 0
